app/models/submission.rb

```diff
@@ -0,0 +1,40 @@
+class Submission < ApplicationRecord
+  belongs_to :user
+
+  has_many :comments, class_name: "Submission", foreign_key: :post_id
+  belongs_to :post, class_name: "Submission", optional: true
+
+  validates :user, presence: true
+  validates :title, :url, presence: true, if: :url_submission?
+
+  validate :text_submission
+  validate :at_least_one_field_present
+
+  def limit_text
+    ary = self.text.chars.shift(80)
+    shift_pos = 0
+    ary.reverse.each_with_index do |val, index|
+      shift_pos = index + 1 if val.blank?
+      break if shift_pos > 0
+    end
+    "#{ary.shift(ary.count - shift_pos).join}..."
+  end
+
+  private
+
+    def url_submission?
+      url.present? || title.present?
+    end
+
+    def text_submission
+      if text.present? && (url.present? || title.present?)
+        errors.add(:base, "You may submit either text content OR a url link, NOT both")
+      end
+    end
+
+    def at_least_one_field_present
+      if text.blank? && url.blank? && title.blank?
```

Comment:
Why `blank?` here, but `present?` above?

Reply:
Good question - the method above, `text_submission`, is designed to say, if you submit a text post, then the url and title fields should NOT be present, so a message is added to the errors object. This method, `at_least_one_field_present` is saying, if all of the fields are blank, add to the errors object. 

Do you think the following would be more readable for `text_submission`? 
`if text.present? && (!url.blank? || !title.blank?)`

Reply:
Sorry. I left a bad comment. `blank?` is the opposite of `.present?`. Sorry :(

Reply:
no worries

Reply:
Fun Ruby trick: `if [text, url, title].all?(&:blank?)`